pyracf/common/logger.py

```diff
@@ -93,33 +95,63 @@ def log_debug(self, header_message: str, message: str) -> None:
         )
         print(f"{header}\n{message}")
 
-    def redact_strings(
+    def __redact_request_dictionary(
         self,
-        string: Union[str, bytes],
-        redact_strings: List[str],
-    ) -> str:
-        """Redact a list of strings or sequences of bytes in a string or bytes object"""
-        if not redact_strings:
-            return string
-        for to_redact in redact_strings:
-            if isinstance(string, bytes):
-                string = self.redact_string(
-                    string, redact_string=bytes(to_redact, "utf-8")
-                )
-            else:
-                string = self.redact_string(string, redact_string=to_redact)
-        return string
+        dictionary: dict,
+        secret_traits: dict,
+    ) -> dict:
+        """Redact a list of specified secret traits in a dictionary"""
+        for seg_trait in secret_traits:
+            segment = seg_trait.split(":")[0]
+            try:
+                if not dictionary.get(segment, {}).get(seg_trait, {}).get("value", ""):
+                    continue
+                dictionary[segment][seg_trait]["value"] = "********"
+            except AttributeError:
```

Comment:
I am guessing that the exception handler here is for line `109`

```python
               dictionary[segment][seg_trait]["value"] = "********"
```

In that case, you should be able to limit the scope of that exception handler to just that line. Also, it is better practice to ignore an exception by using `pass`. I think you should only use `continue` if there is code after that try except block that you want to skip. Also, changing `dictionary.get(segment, {}).get(seg_trait, {}).get("value", "")` to `dictionary.get(segment, {}).get(seg_trait, {}).get("value", None)` is clearer in my opinion.

```python
    ) -> dict:
        """Redact a list of specified secret traits in a dictionary"""
        for seg_trait in secret_traits:
            segment = seg_trait.split(":")[0]
            if not dictionary.get(segment, {}).get(seg_trait, {}).get("value", None):
                continue
            try:
                dictionary[segment][seg_trait]["value"] = "********"
            except AttributeError:
                pass
        return dictionary
```

Reply:
Actually, the attribute error had been for the "if" statement. In testing (I think for extract requests), I was hitting Attribute Errors for requests since their segments dictionary leads to bools not more dictionaries. The existing solution was clunky, and should have been better optimized. I replaced it with the following and don't see the error in unit testing.

```python
    ) -> dict:
        """Redact a list of specified secret traits in a dictionary"""
        for seg_trait in secret_traits:
            segment = seg_trait.split(":")[0]
            if not isinstance(dictionary.get(segment, {}).get(seg_trait, {}),dict):
                return dictionary
            if not dictionary.get(segment, {}).get(seg_trait, {}).get("value", ""):
                continue
            dictionary[segment][seg_trait]["value"] = "********"
        return dictionary
```

Reply:
This doesn't seem quite right. You don't want to return prematurely. I think something like this should work:

```python
    ) -> dict:
        """Redact a list of specified secret traits in a dictionary"""
        for seg_trait in secret_traits:
            segment = seg_trait.split(":")[0]
            if dictionary.get(segment, {}).get(seg_trait, {}).get("value", None):
                continue
            dictionary[segment][seg_trait]["value"] = "********"
        return dictionary
```

If the value doesn't exist `continue` and don't try to redact anything, if you do find the value, redact it.

Reply:
The intent behind that first `if` statement is that if we meet the condition specified, it is because the traits dictionary is of the form of an extract, where it takes the form { 'segment': bool } instead of  { 'segment': { 'trait': {} } } as expected for other requests. we do not redact extract request dictionaries because there are no fields to redact, so there is no reason to continue.

With that in mind, your change did work with a slight correction, so the `get` behavior is not how I initially expected.

```python
    ) -> dict:
        """Redact a list of specified secret traits in a request dictionary"""
        for seg_trait in secret_traits:
            segment = seg_trait.split(":")[0]
            if not dictionary.get(segment, {}).get(seg_trait, {}).get("value", None):
                continue
            dictionary[segment][seg_trait]["value"] = "********"
        return dictionary
```

Reply:
I agree. Even though this technically gets used with profile extract, this logic is cleaner and easier to follow. Even though this logic is pointless for profile extract, the amount of cycles that get wasted running through this loop is negligible. My only suggestion would be to maybe document this nuance in the docstring for this function so that there is no confusion about what happens with profile extract.

Reply:
Changed docstring to:

```python
        """
        Redact a list of specified secret traits in a request dictionary.
        While this flow is technically used in extract functions, there are no traits to redact.
        """
```